graylog2-server/src/main/java/org/graylog2/inputs/persistence/InputStatusService.java

```diff
@@ -0,0 +1,93 @@
+package org.graylog2.inputs.persistence;
+
+import com.google.common.eventbus.EventBus;
+import com.mongodb.DB;
+import com.mongodb.DBCollection;
+import org.apache.shiro.event.Subscribe;
+import org.bson.types.ObjectId;
+import org.graylog2.bindings.providers.MongoJackObjectMapperProvider;
+import org.graylog2.database.MongoConnection;
+import org.graylog2.rest.models.system.inputs.responses.InputDeleted;
+import org.mongojack.JacksonDBCollection;
+import org.mongojack.WriteResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.util.Optional;
+
+/**
+ * This is a simple wrapper around MongoDB to allow storage of state data for stateful Inputs.
+ *
+ * Inputs using this service are responsible for defining their own model for InputStatusRecord.inputStateData
+ */
+public class InputStatusService {
+    private static final Logger LOG = LoggerFactory.getLogger(InputStatusService.class);
+
+    private static final String INPUT_STATUS_COLLECTION = "input_status";
+
+    private final JacksonDBCollection<InputStatusRecord, ObjectId> statusCollection;
+
+    @Inject
+    public InputStatusService(MongoConnection mongoConnection,
+                              MongoJackObjectMapperProvider objectMapperProvider,
+                              EventBus eventBus) {
+        DB mongoDatabase = mongoConnection.getDatabase();
+        DBCollection collection = mongoDatabase.getCollection(INPUT_STATUS_COLLECTION);
+
+        eventBus.register(this);
```

Comment:
I think we should bind this service in a singleton scope (`bind(InputStatusService.class).asEagerSingleton()` and add `@Singleton` annotation to this class for documentation purposes) in the guice bindings to avoid multiple instances of this class. The problem is that every instance would register with the `EventBus` and since we don't have lifecycle support for this object where we could call `unregister` on shutdown, we would basically leak references into the event bus. We have that issue in other services as well and mostly solve it by binding the service as a singleton.

Reply:
Will do